### PR TITLE
fix(server): [YW-267] atomic auto-draft dedup — close TOCTOU race

### PR DESCRIPTION
**Pull request by youhaowei**
## Summary

- Wraps the `createInsight` check-and-insert in a single `ctx.db.transaction` so two concurrent calls for the same `baseTableId` always converge on one unmodified draft — no TOCTOU window.
- Removes the client-side read-then-write dedup gate (`if existingDraft navigate + return`) from `createInsightFromTable`; the client now reads existing insights only to compute a gap-free numeric suffix for the modified-insight name, then always delegates to the server.
- Adds `skipDedup: true` option for `createInsightFromInsight` callers so derived insights are never silently rerouted to an existing unmodified draft for the same `baseTableId`.

Tracked internally as YW-267.

## Test plan

- [x] `turbo typecheck` — 44/44 tasks pass
- [x] `turbo lint` — clean
- [x] `prettier --check` — clean
- [x] `apps/server` app-artifacts tests — 9/9 pass (4 new dedup contract tests + 1 skipDedup test)
- [x] `packages/app` useCreateInsight tests — 33/33 pass (TOCTOU simulation, derived-insight skipDedup contract)
- [x] Key concurrency test: `Promise.all` fires two `createInsight` calls for the same `baseTableId` without awaiting the first; both resolve to the same `id` and exactly one row in the DB

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details open><summary><h3>Greptile Summary</h3></summary>

This PR closes the TOCTOU race in auto-draft deduplication by moving the check-and-insert into a single server-side `ctx.db.transaction`, and replaces the now-redundant client-side early-return gate with a pure UX read (suffix computation only). The `isUnmodifiedDraft` predicate is extracted to `@dashframe/types` as a single source of truth shared by both the renderer hook and the server gate.

- **Atomic server dedup**: `createInsight` wraps the look-up + insert in one transaction; concurrent auto-draft calls for the same `baseTableId` converge on one row via `reuseUnmodifiedDraft: true` (opt-in, default is always-insert).
- **Client dedup gate removed**: `createInsightFromTable` now reads existing insights only to compute gap-free numeric suffixes for the modified-insight path; it no longer short-circuits before reaching the server.
- **Shared predicate**: `isUnmodifiedDraft` / `InsightDraftShape` moved to `@dashframe/types`, eliminating the previously duplicated client/server logic.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge. The transaction wrapping is correct for the single-connection PGLite target and the limitations for multi-connection stores are explicitly documented with a revisit trigger.

The change is well-scoped: the server transaction closes the specific TOCTOU window it targets, the client-side gate removal is backed by the server taking over responsibility, the shared isUnmodifiedDraft predicate eliminates the previously noted drift risk, and the opt-in semantics (reuseUnmodifiedDraft defaults to always-insert) mean callers that omit the flag can never accidentally trigger the dedup path. Test coverage is thorough across sequential, concurrent, and edge-case scenarios.

No files require special attention.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/server/src/functions/app-artifacts.ts | Core fix: wraps check-and-insert in a db transaction with opt-in reuseUnmodifiedDraft; includes inline comments for multi-connection and full-scan caveats. |
| packages/types/src/insights.ts | Adds shared InsightDraftShape interface and isUnmodifiedDraft function; adds reuseUnmodifiedDraft opt-in to InsightMutations.create options. |
| packages/app/src/hooks/useCreateInsight.ts | Removes client-side early-return dedup gate; now reads insights only to compute suffix names; delegates all dedup to the server via reuseUnmodifiedDraft. |
| apps/server/src/functions/app-artifacts.test.ts | Adds 6 new dedup-contract tests (sequential reuse, TOCTOU simulation, modified-draft bypass, pre-populated bypass, no-reuse-flag default, explicit reuse=false) using real PGLite. |
| packages/app/src/hooks/useCreateInsight.test.tsx | Updates existing dedup test assertions and adds a TOCTOU concurrent-call convergence test. |
| packages/types/src/index.ts | Exports new InsightDraftShape type and isUnmodifiedDraft function from the types package barrel. |
| packages/app-data/src/insights.ts | Threads reuseUnmodifiedDraft through the useInsightMutations create hook to match the updated InsightMutations interface. |

</details>

</details>

<sub>Reviews (3): Last reviewed commit: ["docs(insight): record single-connection ..."](https://github.com/youhaowei/dashframe/commit/9231c4d771f0f00c06cdc6b4ca4e1a0f36bd8511) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38047325)</sub>

### CODE REVIEW

File: apps/server/src/functions/app-artifacts.ts [hunk 1]
```diff
@@ -768,18 +787,57 @@ const createInsight = mutation({
...
+          .filter(
+            (r) =>
+              (r.definition as InsightDefinition).baseTableId === baseTableId,
+          )
+          .find((r) => isUnmodifiedDraft(r.definition as InsightDefinition));
+
+        if (existingDraft) {
+          return { id: existingDraft.id };
+        }
+      }
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/app-artifacts.ts:843]: Addressed in commit 25c5a66f. The full-table-scan is now documented at the `tx.from(insights).all()` call site with the exact mitigation you suggested: a NOTE explaining that `baseTableId` lives in the unindexed JSONB `definition` column (no JSONB-path filtering in @wystack/db), plus a trigger to promote it to a top-level indexed column or add a JSONB expression index when insight-table scale shows up in latency. Correctness is sound at current scale; this is a deliberate LATER with a recorded trigger.

File: apps/server/src/functions/app-artifacts.ts [hunk 2]
```diff
@@ -768,18 +787,57 @@ const createInsight = mutation({
...
+        (opts.metrics?.length ?? 0) === 0;
+
+      if (incomingIsUnmodified) {
+        const rows = (await tx.from(insights).all()) as InsightRow[];
+        const existingDraft = rows
+          .filter(
+            (r) =>
+              (r.definition as InsightDefinition).baseTableId === baseTableId,
+          )
+          .find((r) => isUnmodifiedDraft(r.definition as InsightDefinition));
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/app-artifacts.ts:838]: Same full-scan point — addressed in 25c5a66f with a NOTE + scale trigger at the call site. JSONB-path filtering (`WHERE definition->>'baseTableId' = $1`) is not available in @wystack/db today (operators only support flat top-level columns), so the JS filter is unavoidable until baseTableId is promoted to a top-level indexed column. Recorded as a LATER trigger rather than a bare scan.

File: packages/types/src/insights.ts [hunk 3]
```diff
@@ -132,6 +132,10 @@ export interface InsightMutations {
     options?: {
       selectedFields?: UUID[];
       metrics?: InsightMetric[];
+      /** When true, bypass auto-draft dedup and always insert a new row.
+       *  Used by createInsightFromInsight so derived insights are never
+       *  silently rerouted to an existing unmodified draft. */
+      skipDedup?: boolean;
     },
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/types/src/insights.ts]: Reworked. `skipDedup` is removed entirely. Dedup is now opt-IN via `reuseUnmodifiedDraft` (default false → create means create), so no internal mechanism is leaked: the auto-draft path opts in, the derived path simply omits it. The flag describes a caller intent ("reuse an existing unmodified draft for this table if one exists"), not a server escape hatch — a future caller that omits it gets the safe default (always insert).
